JCache is leaking memory

There is a MaxSizePolicy set to 10,000 entries. Hence in theory this snippet should run indefinitely. 

```
    @Test
    public void testEviction() throws Exception {
        CacheManager cacheManager = cachingProvider1.getCacheManager();
        String cacheName = "myCacheName";
        CacheConfig<Long, Long> config = new CacheConfig<Long, Long>();
        final ICache<Long, Long> cache = (ICache<Long, Long>) cacheManager.createCache(cacheName, config);

        ExpiryPolicy expiryPolicy = new CreatedExpiryPolicy(Duration.ONE_MINUTE);
        List<ICompletableFuture<Void>> futures = new ArrayList<ICompletableFuture<Void>>();
        for (long i = 0; ; i++) {
            ICompletableFuture<Void> future = cache.putAsync(i, 0l, expiryPolicy);
            futures.add(future);
            if (i % 10000 == 0) {
                for (ICompletableFuture<Void> f : futures) {
                    f.get();
                }
                System.out.println("At iteration " + i + ", cache size: " + cache.size());
                futures.clear();
            }
        }
    }
```

I can see there is an upper bound on cache size as expected, however it's still leaking memory. 

```
2:05:19,180  INFO [HealthMonitor] hz._hzInstance_2_dev.HealthMonitor - [127.0.0.1]:5702 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=148.1M, swap.space.total=15.7G, swap.space.free=13.4G, heap.memory.used=3.5G, heap.memory.free=240.6M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=93.67%, heap.memory.used/max=93.67%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=155, major.gc.time=1057856ms, load.process=82.00%, load.system=97.00%, load.systemAverage=489.00%, thread.count=67, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=1285, executor.q.priorityOperation.size=0, executor.q.response.size=0, operations.remote.size=0, operations.running.size=0, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
12:05:19,182  INFO [HealthMonitor] hz._hzInstance_1_dev.HealthMonitor - [127.0.0.1]:5701 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=151.0M, swap.space.total=15.7G, swap.space.free=13.4G, heap.memory.used=3.5G, heap.memory.free=239.7M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=93.69%, heap.memory.used/max=93.69%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=155, major.gc.time=1057856ms, load.process=0.00%, load.system=0.00%, load.systemAverage=489.00%, thread.count=67, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=371, executor.q.priorityOperation.size=2, executor.q.response.size=0, operations.remote.size=2800, operations.running.size=2, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
At iteration 18510000 cache size: 14905
At iteration 18520000 cache size: 14905
At iteration 18530000 cache size: 14905
At iteration 18540000 cache size: 14905
At iteration 18550000 cache size: 14905
At iteration 18560000 cache size: 14905
12:05:51,971  INFO [HealthMonitor] hz._hzInstance_1_dev.HealthMonitor - [127.0.0.1]:5701 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=167.5M, swap.space.total=15.7G, swap.space.free=13.4G, heap.memory.used=3.5G, heap.memory.free=256.9M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=93.24%, heap.memory.used/max=93.24%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=161, major.gc.time=1090005ms, load.process=0.00%, load.system=97.00%, load.systemAverage=469.00%, thread.count=66, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=116, executor.q.priorityOperation.size=0, executor.q.response.size=0, operations.remote.size=2293, operations.running.size=1, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
12:05:51,972  INFO [HealthMonitor] hz._hzInstance_2_dev.HealthMonitor - [127.0.0.1]:5702 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=167.5M, swap.space.total=15.7G, swap.space.free=13.4G, heap.memory.used=3.5G, heap.memory.free=256.9M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=93.24%, heap.memory.used/max=93.24%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=161, major.gc.time=1090005ms, load.process=83.00%, load.system=100.00%, load.systemAverage=469.00%, thread.count=66, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=1915, executor.q.priorityOperation.size=0, executor.q.response.size=0, operations.remote.size=0, operations.running.size=0, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
At iteration 18570000 cache size: 14905
At iteration 18580000 cache size: 14905
At iteration 18590000 cache size: 14905
At iteration 18600000 cache size: 14905
At iteration 18610000 cache size: 14905
12:06:25,012  INFO [HealthMonitor] hz._hzInstance_1_dev.HealthMonitor - [127.0.0.1]:5701 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=166.3M, swap.space.total=15.7G, swap.space.free=13.4G, heap.memory.used=3.5G, heap.memory.free=240.3M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=93.68%, heap.memory.used/max=93.68%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=167, major.gc.time=1122221ms, load.process=83.00%, load.system=96.00%, load.systemAverage=490.00%, thread.count=65, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=437, executor.q.priorityOperation.size=97, executor.q.response.size=0, operations.remote.size=1508, operations.running.size=1, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
12:06:25,012  INFO [HealthMonitor] hz._hzInstance_2_dev.HealthMonitor - [127.0.0.1]:5702 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=166.3M, swap.space.total=15.7G, swap.space.free=13.4G, heap.memory.used=3.5G, heap.memory.free=239.8M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=93.69%, heap.memory.used/max=93.69%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=167, major.gc.time=1122221ms, load.process=0.00%, load.system=100.00%, load.systemAverage=490.00%, thread.count=65, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=3, executor.q.priorityOperation.size=0, executor.q.response.size=0, operations.remote.size=0, operations.running.size=0, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
At iteration 18620000 cache size: 14905
At iteration 18630000 cache size: 14905
At iteration 18640000 cache size: 14905
At iteration 18650000 cache size: 14905
[...]
12:19:54,962  INFO [HealthMonitor] hz._hzInstance_1_dev.HealthMonitor - [127.0.0.1]:5701 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=161.7M, swap.space.total=15.7G, swap.space.free=13.0G, heap.memory.used=3.6G, heap.memory.free=147.2M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=96.13%, heap.memory.used/max=96.13%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=307, major.gc.time=1923846ms, load.process=0.00%, load.system=97.00%, load.systemAverage=511.00%, thread.count=66, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=582, executor.q.priorityOperation.size=0, executor.q.response.size=0, operations.remote.size=2435, operations.running.size=1, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
12:19:54,963  INFO [HealthMonitor] hz._hzInstance_2_dev.HealthMonitor - [127.0.0.1]:5702 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=161.7M, swap.space.total=15.7G, swap.space.free=13.0G, heap.memory.used=3.6G, heap.memory.free=147.2M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=96.13%, heap.memory.used/max=96.13%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=307, major.gc.time=1923846ms, load.process=84.00%, load.system=0.00%, load.systemAverage=511.00%, thread.count=66, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=1, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=387, executor.q.priorityOperation.size=0, executor.q.response.size=0, operations.remote.size=0, operations.running.size=0, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
At iteration 19150000 cache size: 14905
12:20:29,387  INFO [HealthMonitor] hz._hzInstance_2_dev.HealthMonitor - [127.0.0.1]:5702 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=161.8M, swap.space.total=15.7G, swap.space.free=13.0G, heap.memory.used=3.6G, heap.memory.free=144.7M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=96.19%, heap.memory.used/max=96.19%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=313, major.gc.time=1958186ms, load.process=0.00%, load.system=97.00%, load.systemAverage=558.00%, thread.count=66, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=3412, executor.q.priorityOperation.size=0, executor.q.response.size=0, operations.remote.size=0, operations.running.size=1, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
12:20:29,388  INFO [HealthMonitor] hz._hzInstance_1_dev.HealthMonitor - [127.0.0.1]:5701 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=161.8M, swap.space.total=15.7G, swap.space.free=13.0G, heap.memory.used=3.6G, heap.memory.free=144.7M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=96.19%, heap.memory.used/max=96.19%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=313, major.gc.time=1958186ms, load.process=82.00%, load.system=100.00%, load.systemAverage=558.00%, thread.count=66, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=237, executor.q.priorityOperation.size=0, executor.q.response.size=0, operations.remote.size=2510, operations.running.size=0, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
At iteration 19160000 cache size: 14905
12:21:02,344  INFO [HealthMonitor] hz._hzInstance_2_dev.HealthMonitor - [127.0.0.1]:5702 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=188.3M, swap.space.total=15.7G, swap.space.free=13.0G, heap.memory.used=3.6G, heap.memory.free=143.6M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=96.22%, heap.memory.used/max=96.22%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=319, major.gc.time=1991065ms, load.process=85.00%, load.system=97.00%, load.systemAverage=529.00%, thread.count=66, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=223, executor.q.priorityOperation.size=0, executor.q.response.size=0, operations.remote.size=0, operations.running.size=0, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
12:21:07,832  INFO [HealthMonitor] hz._hzInstance_1_dev.HealthMonitor - [127.0.0.1]:5701 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=188.1M, swap.space.total=15.7G, swap.space.free=13.0G, heap.memory.used=3.6G, heap.memory.free=143.2M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=96.23%, heap.memory.used/max=96.23%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=320, major.gc.time=1996540ms, load.process=85.00%, load.system=97.00%, load.systemAverage=518.00%, thread.count=66, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=281, executor.q.priorityOperation.size=0, executor.q.response.size=0, operations.remote.size=3342, operations.running.size=2, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
12:21:33,007  INFO [HealthMonitor] hz._hzInstance_2_dev.HealthMonitor - [127.0.0.1]:5702 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=148.6M, swap.space.total=15.7G, swap.space.free=13.0G, heap.memory.used=3.6G, heap.memory.free=141.3M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=96.28%, heap.memory.used/max=96.28%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=324, major.gc.time=2021669ms, load.process=82.00%, load.system=97.00%, load.systemAverage=491.00%, thread.count=66, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=4676, executor.q.priorityOperation.size=0, executor.q.response.size=0, operations.remote.size=0, operations.running.size=0, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
12:21:41,403 FATAL [RemoteOperationExceptionHandler] hz._hzInstance_2_dev.partition-operation.thread-2 - Problem while reading DataSerializable, namespace: -1, id: 1, class: 'null', exception: java.lang.reflect.InvocationTargetException
com.hazelcast.nio.serialization.HazelcastSerializationException: Problem while reading DataSerializable, namespace: -1, id: 1, class: 'null', exception: java.lang.reflect.InvocationTargetException
    at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:120)
    at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:1)
    at com.hazelcast.nio.serialization.StreamSerializerAdapter.toObject(StreamSerializerAdapter.java:65)
    at com.hazelcast.nio.serialization.SerializationServiceImpl.toObject(SerializationServiceImpl.java:260)
    at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:186)
    at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.loadOperation(BasicOperationService.java:638)
    at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.handle(BasicOperationService.java:621)
    at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.access$0(BasicOperationService.java:619)
    at com.hazelcast.spi.impl.BasicOperationService$BasicDispatcherImpl.dispatch(BasicOperationService.java:566)
    at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.process(BasicOperationScheduler.java:466)
    at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.doRun(BasicOperationScheduler.java:458)
    at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.run(BasicOperationScheduler.java:432)
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
    at com.hazelcast.nio.UTFEncoderDecoder$FastStringCreator.buildString(UTFEncoderDecoder.java:786)
    at com.hazelcast.nio.UTFEncoderDecoder.readUTF0(UTFEncoderDecoder.java:516)
    at com.hazelcast.nio.UTFEncoderDecoder.readUTF(UTFEncoderDecoder.java:77)
    at com.hazelcast.nio.serialization.ByteArrayObjectDataInput.readUTF(ByteArrayObjectDataInput.java:434)
    at com.hazelcast.spi.Operation.readData(Operation.java:290)
    at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:111)
    ... 11 more
Caused by: java.lang.reflect.InvocationTargetException
    at sun.reflect.GeneratedConstructorAccessor1.newInstance(Unknown Source)
    at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
    at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
    at com.hazelcast.nio.UTFEncoderDecoder$FastStringCreator.buildString(UTFEncoderDecoder.java:781)
    ... 16 more
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
    at java.util.Arrays.copyOfRange(Arrays.java:2694)
    at java.lang.String.<init>(String.java:203)
    at java.lang.String.<init>(String.java:611)
    ... 20 more
12:21:41,405  INFO [HealthMonitor] hz._hzInstance_1_dev.HealthMonitor - [127.0.0.1]:5701 [dev] [3.4-RC1-SNAPSHOT] processors=4, physical.memory.total=15.4G, physical.memory.free=165.2M, swap.space.total=15.7G, swap.space.free=13.0G, heap.memory.used=3.6G, heap.memory.free=142.2M, heap.memory.total=3.7G, heap.memory.max=3.7G, heap.memory.used/total=96.26%, heap.memory.used/max=96.26%, minor.gc.count=208, minor.gc.time=25563ms, major.gc.count=325, major.gc.time=2030047ms, load.process=86.00%, load.system=97.00%, load.systemAverage=493.00%, thread.count=66, thread.peakCount=71, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operation.size=110, executor.q.priorityOperation.size=0, executor.q.response.size=0, operations.remote.size=1763, operations.running.size=0, proxy.count=1, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
12:21:49,272 FATAL [OperationService] hz._hzInstance_1_dev.IO.thread-in-0 - [127.0.0.1]:5701 [dev] [3.4-RC1-SNAPSHOT] While processing response...
com.hazelcast.core.HazelcastException: No invocation for response: NormalResponse{callId=19166233, urgent=true, value=com.hazelcast.nio.serialization.HazelcastSerializationException: Problem while reading DataSerializable, namespace: -1, id: 1, class: 'null', exception: java.lang.reflect.InvocationTargetException, backupCount=0}
    at com.hazelcast.spi.impl.BasicOperationService$ResponsePacketHandler.notifyRemoteCall(BasicOperationService.java:602)
    at com.hazelcast.spi.impl.BasicOperationService$ResponsePacketHandler.handle(BasicOperationService.java:586)
    at com.hazelcast.spi.impl.BasicOperationService$ResponsePacketHandler.access$0(BasicOperationService.java:580)
    at com.hazelcast.spi.impl.BasicOperationService$BasicDispatcherImpl.dispatch(BasicOperationService.java:564)
    at com.hazelcast.spi.impl.BasicOperationScheduler$ResponseThread.process(BasicOperationScheduler.java:532)
    at com.hazelcast.spi.impl.BasicOperationScheduler$ResponseThread.access$1(BasicOperationScheduler.java:529)
    at com.hazelcast.spi.impl.BasicOperationScheduler.execute(BasicOperationScheduler.java:277)
    at com.hazelcast.spi.impl.BasicOperationService.executeOperation(BasicOperationService.java:241)
    at com.hazelcast.spi.impl.NodeEngineImpl.handlePacket(NodeEngineImpl.java:280)
    at com.hazelcast.nio.NodeIOService.handleMemberPacket(NodeIOService.java:107)
    at com.hazelcast.nio.tcp.DefaultPacketReader.handlePacket(DefaultPacketReader.java:60)
    at com.hazelcast.nio.tcp.DefaultPacketReader.readPacket(DefaultPacketReader.java:47)
    at com.hazelcast.nio.tcp.SocketPacketReader.read(SocketPacketReader.java:41)
    at com.hazelcast.nio.tcp.ReadHandler.handle(ReadHandler.java:79)
    at com.hazelcast.nio.tcp.InSelectorImpl.handleSelectionKey(InSelectorImpl.java:49)
    at com.hazelcast.nio.tcp.AbstractIOSelector.handleSelectionKeys(AbstractIOSelector.java:159)
    at com.hazelcast.nio.tcp.AbstractIOSelector.run(AbstractIOSelector.java:132)
12:21:49,276 FATAL [OperationService] hz._hzInstance_2_dev.partition-operation.thread-2 - [127.0.0.1]:5702 [dev] [3.4-RC1-SNAPSHOT] Problem while reading DataSerializable, namespace: -1, id: 1, class: 'null', exception: java.lang.reflect.InvocationTargetException
com.hazelcast.nio.serialization.HazelcastSerializationException: Problem while reading DataSerializable, namespace: -1, id: 1, class: 'null', exception: java.lang.reflect.InvocationTargetException
    at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:120)
    at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:1)
    at com.hazelcast.nio.serialization.StreamSerializerAdapter.toObject(StreamSerializerAdapter.java:65)
    at com.hazelcast.nio.serialization.SerializationServiceImpl.toObject(SerializationServiceImpl.java:260)
    at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:186)
    at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.loadOperation(BasicOperationService.java:638)
    at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.handle(BasicOperationService.java:621)
    at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.access$0(BasicOperationService.java:619)
    at com.hazelcast.spi.impl.BasicOperationService$BasicDispatcherImpl.dispatch(BasicOperationService.java:566)
    at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.process(BasicOperationScheduler.java:466)
    at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.doRun(BasicOperationScheduler.java:458)
    at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.run(BasicOperationScheduler.java:432)
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
    at com.hazelcast.nio.UTFEncoderDecoder$FastStringCreator.buildString(UTFEncoderDecoder.java:786)
    at com.hazelcast.nio.UTFEncoderDecoder.readUTF0(UTFEncoderDecoder.java:516)
    at com.hazelcast.nio.UTFEncoderDecoder.readUTF(UTFEncoderDecoder.java:77)
    at com.hazelcast.nio.serialization.ByteArrayObjectDataInput.readUTF(ByteArrayObjectDataInput.java:434)
    at com.hazelcast.spi.Operation.readData(Operation.java:290)
    at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:111)
    ... 11 more
Caused by: java.lang.reflect.InvocationTargetException
    at sun.reflect.GeneratedConstructorAccessor1.newInstance(Unknown Source)
    at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
    at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
    at com.hazelcast.nio.UTFEncoderDecoder$FastStringCreator.buildString(UTFEncoderDecoder.java:781)
    ... 16 more
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
    at java.util.Arrays.copyOfRange(Arrays.java:2694)
    at java.lang.String.<init>(String.java:203)
    at java.lang.String.<init>(String.java:611)
    ... 20 more
12:22:05,754  INFO [HealthMonitor] hz._hzInstance_2_dev.HealthMon
```

![image](https://cloud.githubusercontent.com/assets/158619/5411899/bddf69f6-81fe-11e4-9a55-b3c17f35b71d.png)

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheRecordStore.java
@@ -723,6 +723,14 @@ public abstract class AbstractCacheRecordStore<R extends CacheRecord, CRM extend
     }
 
     @Override
+    public void putRecord(Data key, CacheRecord record) {
+        if (!records.containsKey(key)) {
+            evictIfRequired();
+        }
+        records.put(key, (R) record);
+    }
+
+    @Override
     public CacheRecord removeRecord(Data key) {
         return records.remove(key);
     }

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/ICacheRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/ICacheRecordStore.java
@@ -344,12 +344,23 @@ public interface ICacheRecordStore {
     /**
      * Associates the specified record with the specified key.
      * This is simply a put operation on the internal map data
-     * without any CacheLoad.
+     * without any CacheLoad. It also <b>DOES NOT</b> trigger eviction,
+     * be aware of the fact it might cause an OutOfMemoryException!
      *
      * @param key the key to the entry.
      * @param record the value to be associated with the specified key.
      */
     void setRecord(Data key, CacheRecord record);
+
+    /**
+     * Associates the specified record with the specified key.
+     * This is simply a put operation on the internal map data
+     * without any CacheLoad. It also <b>DOES</b> trigger eviction!
+     *
+     * @param key the key to the entry.
+     * @param record the value to be associated with the specified key.
+     */
+    void putRecord(Data key, CacheRecord record);
 
     /**
      * Removes the record for a key.

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CachePutAllBackupOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CachePutAllBackupOperation.java
@@ -63,7 +63,7 @@ public class CachePutAllBackupOperation
             throws Exception {
         if (cacheRecords != null) {
             for (Map.Entry<Data, CacheRecord> entry : cacheRecords.entrySet()) {
-                cache.setRecord(entry.getKey(), entry.getValue());
+                cache.putRecord(entry.getKey(), entry.getValue());
             }
         }
     }

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CachePutBackupOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CachePutBackupOperation.java
@@ -54,7 +54,7 @@ public class CachePutBackupOperation
             throws Exception {
         CacheService service = getService();
         ICacheRecordStore cache = service.getOrCreateCache(name, getPartitionId());
-        cache.setRecord(key, cacheRecord);
+        cache.putRecord(key, cacheRecord);
         response = Boolean.TRUE;
     }
 

